server/src/handlers/http/middleware.rs

```diff
@@ -75,25 +76,33 @@ where
 
     forward_ready!(service);
 
-    fn call(&self, req: ServiceRequest) -> Self::Future {
+    fn call(&self, mut req: ServiceRequest) -> Self::Future {
+        let creds = req.extract::<BasicAuth>().into_inner();
+        let creds: Result<(String, String), Error> = creds.map_err(Into::into).and_then(|creds| {
+            let username = creds.user_id().trim().to_owned();
+            if let Some(password) = creds.password() {
+                Ok((username, password.trim().to_owned()))
+            } else {
+                Err(ErrorBadRequest("Password is required in basic auth"))
+            }
+        });
         let stream = if self.match_stream {
             req.match_info().get("logstream")
         } else {
             None
         };
-        let extensions = req.extensions();
-        let username = extensions
-            .get::<String>()
-            .expect("authentication layer verified username");
-        let is_auth = Users.check_permission(username, self.action, stream);
-        drop(extensions);
+        let is_auth = creds.map(|creds| {
+            let (username, password) = creds;
+            Users.authenticate(username, password, self.action, stream)
+        });
```

Comment:
what is `is_auth`?

Reply:
The type is result of authentication and authorization. Type of creds is Result<> which i can only return in the async block down below. So have to `.map` for calling authenticate

`Users.authenticate` does both authentication and authorization.